### PR TITLE
[PLT-959] Browser Compatibility 

### PR DESCRIPTION
**Pull request by moonmeister**
#### Summary
Implements a descriptive error page for browser compatibility. 

I've also cleaned up some error strings that were placed in the backend but never referenced (only by the front end, where they already existed). 

My only major question is do we need to be able to localize the URLs to the browser pages?

#### Ticket Link
#6841 
[jira](https://mattermost.atlassian.net/browse/PLT-959)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Has UI changes
- [x] Includes text changes and localization file

### CODE REVIEW

**Comment by jasonblais** [comment]
@esethna Do we want to mention the Edge versions we support?
@enahum Is the best person on the localized strings.

**Comment by moonmeister** [comment]
I also forgot to mention the fact that while this error message is working when a browser is detected as incompatible the compatiblity check isn't correctly detecting the list provided. It doesn't even check Chrome, FF, or Edge. It only checks IE and Safari. This seems like it should be addressed, but should it be done under this pull request?

**Comment by moonmeister** [comment]
So, I've done some messing with the compatibility check and I have a way to check all browsers and it simplifies the incompatible browser string. I can merge the code into the PR if desired.

That said, I have a question. If someone is truly using an incompatible browser. Will they even be able to render the error page? For instance if they have JS disabled or their browser just doesn't work with React due to its age. Would it be better to create a error page template with a Go template and build the page in go on the server side so all the user get's served is basic HTML and maybe CSS? This way, no matter how outdated the browser or great the incompatibility, the error page should render. 

I'll checkout out why the build failed.

**Comment by moonmeister** [comment]
It looks like the file upload test on the APIv4 tests is failing. Not sure if this is related to a change I made, doesn't seem like like it would be. If anyone has a thought on them matter that'd be appreciated. Thanks.

```
=== RUN   TestUploadFile
[2017/07/15 23:08:55 UTC] [DEBG] POST - /api/v4/files
[2017/07/15 23:08:57 UTC] [DEBG] POST - /api/v4/files
[2017/07/15 23:08:57 UTC] [EROR] /api/v4/files:Permissions code=403 rid=tds11gnjjtyomxx5pg3rga4izw uid=p5knmjxeg3n3mygsqe1w3eao1c ip=127.0.0.1 You do not have the appropriate permissions [details: userId=p5knmjxeg3n3mygsqe1w3eao1c, permission=upload_file]
[2017/07/15 23:08:57 UTC] [DEBG] POST - /api/v4/files
[2017/07/15 23:08:57 UTC] [DEBG] POST - /api/v4/files
[2017/07/15 23:08:57 UTC] [EROR] /api/v4/files:uploadFile code=501 rid=118o9ngggirsie6gi1iz7ktnpo uid=99oaae6m4fnctxz7su4u5s9f4r ip=127.0.0.1 File attachments have been disabled on this server. [details: ]
goroutine 35400 [running]:
runtime/debug.Stack(0x0, 0xc42a7ea6c0, 0x446a88)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/mattermost/platform/api4.CheckNotImplementedStatus(0xc42121cea0, 0xc42a7ea6c0)
	github.com/mattermost/platform/api4/_test/_obj_test/apitestlib.go:668 +0x8f
github.com/mattermost/platform/api4.TestUploadFile(0xc42121cea0)
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/api4/file_test.go:113 +0xe1f
testing.tRunner(0xc42121cea0, 0x13a8d68)
	/usr/local/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:697 +0x2ca
--- FAIL: TestUploadFile (3.36s)
	apitestlib.go:669: actual: 0
	apitestlib.go:670: expected: 501
	apitestlib.go:671: wrong status code
```

**Comment by jasonblais** [comment]
@moonmeister I've rebuilded in case it was a random test failure. Let's see if it's successful now.

Was the test successful locally?

**Comment by moonmeister** [comment]
@jasonblais Yeah, I already rebuilt it already, still failed. No, it' not failing locally.

**Comment by grundleborg** [comment]
@moonmeister @jasonblais the TestUploadFile test is having major problems on Jenkins. Just keep rebuilding it until it passes (there's a ticket elsewhere to fix the test).

**Comment by hmhealey** [comment]
Also, you make a good point about using a static error page for this. We do already have one that's displayed when even JS can't be executed, so that may be enough, but we could definitely consider adding to that, but I think this is fine for now.

**Comment by moonmeister** [comment]
@hmhealey @esethna Two Things:

1. The Edge User Agent reports itself as not version "40" but something closer to 12-14. From what I can see "40" is the edge app version while 12-15 is the edgeHTML render engine. Here are two examples of the Edge user agent. I don't have an actual windows machine to confirm this. Modifying my UA in chrome is failing though. 


- `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.10240`
- `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.79 Safari/537.36 Edge/14.14393`

Side note, if I override my UA string in chrome with the first string it get's reported as Chrome 42. Not sure why it's not being recognized as Edge but it's not. 

2.  If someone is truly using an incompatible browser. Will they even be able to render the error page? For instance if they have JS disabled or their browser just doesn't work with React due to its age. Would it be better to create a error page template with a Go template and build the page in go on the server side so all the user get's served is basic HTML and maybe CSS? This way, no matter how outdated the browser or great the incompatibility, the error page should render. I think it would allow us to simplify some code and make the error page more dynamic.

**Comment by moonmeister** [comment]
@esethna Regarding my previous point 1. In the documentation we probably need to note the app version that is compatible (40+). But for the compatibility check I need to know the edgeHTML version to we support.

**Comment by esethna** [comment]
@moonmeister that would be EdgeHTML v15+

**Comment by esethna** [comment]
@hmhealey any thoughts on point #2?

**Comment by moonmeister** [comment]
@esethna I'll modify it to EdgeHTML 15. Does that mean the minimum Edge version is 39.14942 and not 40+ (https://en.wikipedia.org/wiki/Microsoft_Edge)? Should we just say version 40+ to keep it simple? 

Also, I'm wondering what Go reports the browser as on an actual windows machine. If someone could just `fmt.PRintln(ua.Browser())` in the `CheckBrowserCompatability` function in `web.go` and let me know what it outputs. The reason I'm concerned is while the UA string version for Safari looks like `Safari/537.36` Go reports my Safari browser as `Safari 10.1.1`.  I'm thinking maybe Go will report edge as `Edge 40.xyabc`.

**Comment by hmhealey** [comment]
@esethna I did actually comment above that I didn't think point 2 was necessary right now, but thinking more about it, I think it would be a good idea to do if possible.

@moonmeister Would you be interested in changing this over then to use a static page? For styling, we'd want it to match the current error page, but since this is a one-off error page, it's probably fine to just copy the styles over and apply them directly in the HTML without a stylesheet. We have some code around for parsing and rendering the templates in `utils/html.go` that's used for rendering emails. You can also look at [this really old version of `web/web.go`](https://github.com/mattermost/platform/blob/8e883ad2f516820942cc0131241268c1ea909ff3/web/web.go) where it was used for serve the rest of the webapp

**Comment by moonmeister** [comment]
@hmhealey I'm pretty new to to programming, this is the first/only project I've contributed to and professionally I'm a sysadmin and don't do a lot of programming. Basically, I'm happy to help but my willingness might outpace my skills. I'll take a look and see what I can do.

**Comment by hmhealey** [comment]
Cool, thanks for all your help with this. As usual, feel free to ask if you have any questions. 

If you just want to focus on one part of this at a time, I think your best bet would be to create a very simple static error page template as `templates/unsupported_browser.html`, and then have  `web/web.go` return that when the browser is unsupported. If you take a look [here](https://github.com/mattermost/platform/blob/8e883ad2f516820942cc0131241268c1ea909ff3/web/web.go#L234), returning a template is as easy as instantiating it, optionally setting some props that'll be passed to the template, and then rendering it, like so:
```go
page := NewHtmlTemplatePage("signup_team", c.T("web.root.singup_title"), c.Locale)
page.Props["Info"] = c.T("web.root.singup_info")
page.Render(c, w)
```
That'd be a good start since we're just returning a plain string right now without any formatting right now. After that, you can give the styling a shot, or we can ask @asaadmahmood to help out with that since he does most of the CSS for us.

**Comment by moonmeister** [comment]
@hmhealey I think I was imagining replacing the react error page all together for all errors. Are we on the same page and you're suggesting to start simple or was I not clear?

**Comment by hmhealey** [comment]
I think it's only necessary for the incompatible browser page since it's the only one where we'll ever run into not being able to render the page. For the other errors, I can't see any reason to move those out of React since those browsers should have no problem with it. By keeping those in React, we can share common code, like the styling and localization, and we can more easily control the content of the page

**Comment by moonmeister** [comment]
@hmhealey Sounds good. I think the only thing I can think of is less/simplified code (consolidating error pages into a single location). But I get your point. I'll get started on this when I can, but it probably won't be till next week.

**Comment by moonmeister** [comment]
I've got the template page rendering and am attempting to format it. I'm currently using the existing error page to copy the applied styles into the inline `stlye=""` tag. It's a bit tedious, is there a better way or is this what you were thinking @hmhealey. Or is there a way to include the standard style sheets? Or at least modify the default header to include style info so I'm not doing it line-by-line?

Thanks.

**Comment by hmhealey** [comment]
I'm not sure if there's a way to use the standard style sheets since those are mangled and served up through webpack. I'll see if there's any way to access them without that

**Comment by hmhealey** [comment]
Sounds like that's not possible. To make your life easier though, if you go to the current error page (like `https://pre-release.mattermost.com/error`), you can use the Chrome Devtools to copy the exact HTML and CSS used for the existing error page without having to figure out what CSS rules are being applied

**Comment by moonmeister** [comment]
@hmhealey Okay, I think it's close. 

I found a chrome plugin to extract all the applicable CSS and added it to the header.

I figured out how to make the font awesome icon work. Though it's JS which is kinda defeats the purpose of having this page in go (it looks fine without it). I had to create an "Embed Code" with Font Awesome to get this to work. Not sure if you guys have your own embed code that'd be better to use. Also, should we just remove the ! triangle?

Finally, I'm still not sure browser version detection is working for anything Windows, especially Edge/IE. PLEASE TEST.

**Comment by hmhealey** [comment]
@moonmeister I forgot about that icon. I'll double check with our designer, but I think it's fine to remove it since we don't want to be including code from outside the repository and doing anything else is not worth the effort.

Also, the build is failing on your branch because there's some conflicts with master, so you'll either need to rebase onto master branch or merge it into this one

**Comment by moonmeister** [comment]
@hmhealey Okay, I'll wait to hear on the icon removal. I'll work on the rebase.

**Comment by hmhealey** [comment]
[Here's a static version of the icon](https://pre-release.mattermost.com/files/m5ct1815k7g5dxpmn3zgwi7eph/public?h=F2MkuBq-ODkl2ZxE10CvmreUq84sB9eMhxvMDdPe8bk). If you copy that into `webapp/images`, and then add a line to [this block in `webapp/webpack.config.js`](https://github.com/mattermost/platform/blob/master/webapp/webpack.config.js#L319), it'll make the file available at `/static/images/<image>.png`. In terms of styling, you should use `width: 22px` and `opacity: 0.6` for it

**Comment by esethna** [comment]
Thanks @moonmeister, when you're done with the updates can you please post a screenshot of the error page for UI review?

**Comment by moonmeister** [comment]
Here it is!

screen shot 2017-08-01 at 12 40 28 pm

**Comment by moonmeister** [comment]
@hmhealey @esethna Also, for debugging purposes I put in a line of code `fmt.Println` to log in the terminal what Go sees the browser and browser version as. It's super helpful and should probably be permanent, but I'm guessing this isn't the preferred method for debug logging. So, should it stay or go and if it should stay, how would "correctly" log this?

**Comment by hmhealey** [comment]
We use log4go for all of our debug logging currently. You can see it on line 21 of web.go where we use `l4g.Debug(format string, ...args)`. That line is running the output through the localization code, but you don't need to worry about that with your stuff.

**Comment by moonmeister** [comment]
@hmhealey @esethna I was able to connect to the test server and confirm that the Microsoft Edge detection is not working correctly. I connected with Edge 38.xxxx (EdgeHTML 14.xxx) and it did not throw the error. I'm not able to see the logs though on the test server to know what it did detect the the browser version as.

**Comment by jasonblais** [comment]
@moonmeister You can use the "Test account 1" credentials to access the server logs (that account is a System Admin on the test server)

**Comment by moonmeister** [comment]
@jasonblais @hmhealey Can someone rebuild the test server. I correctly setup logging after it was created so I can't see browser data in the logs. Thanks. Also, this might need to be done a couple times, is it something that's possible for me.

**Comment by moonmeister** [comment]
@hmhealey So Edge 38.xxxx (EdgeHTML 14.xxx) is being detected as Chrome 51. Not sure what we can do about this. Are we sure this version of Edge is incompatible?

**Comment by moonmeister** [comment]
@hmhealey The internet is recommending something like https://modernizr.com to detect if specific features are unavailable and to work around them or then throw an incompatibility error.

**Comment by jasonblais** [comment]
@hmhealey @esethna what's the status of this PR? 

@moonmeister As for Edge compatibility: Mattermost does load on Edge, but there are various bugs on that version, which is why we recommend upgrading to the later version (I think v40?)

**Comment by esethna** [comment]
@hmhealey any recommendations to help detect the Edge version correctly?

**Comment by hmhealey** [comment]
I don't know of any better ways to detect the browser unless we want to submit a PR to https://github.com/mssola/user_agent. I think it's fine if Edge's version isn't detected correctly by this error page since it has such a low market share (<2% based on [this](http://gs.statcounter.com/)) and because Microsoft is pushing so hard to keep users up to date with it, similar to Chrome

**Comment by moonmeister** [comment]
@jasonblais @hmhealey @esethna I don't have a preference. Though it would be nice to be able to detect edge. Looks like there is code in that repo so this is probably a bug. We could merge as is and file a bug plus open an issue on our side to test once the bug is fixed in that project. Whatever ya'll decide just let me know. I think the code is done(short of and revisions requested) unless we decide to wait to fix the Edge issue.

**Comment by moonmeister** [comment]
I went ahead and submitted an Issue https://github.com/mssola/user_agent/issues/36.

**Comment by moonmeister** [comment]
@jasonblais @hmhealey I was just thinking is overly obtrusive to deny a user access to the entire Mattermost instance because of potentially small bug? Especially if they don't have access to another browser for whatever reason (internet cafe, quick task, etc).  Is it reasonable to allow them to "bypass" the error screen if there browser only has a "minor" incompatibility? Maybe if they choose to bypass we set a per session variable that bypasses future warnings but should they start a new session they get warned again. It's annoying enough to get someone to upgrade if they can, but not so intrusive that it hinders those who genuinely can not or don't want to for whatever reason. Just a thought let me know what you think.

**Comment by hmhealey** [comment]
I don't think it's necessary to support that case. Most modern browsers force you to keep up to date, so most users won't hit this anyway. If they do, there's potential security risks opened up by allowing vulnerable browsers to access Mattermost.

Also, your code looks good. Would you be able to add some unit tests for the updated browser detection stuff? If you change `CheckBrowserCompatability` to just take the user agent string, you can add a `TestCheckBrowserCompatability` function to `web/web_test.go` that'll be run with the unit tests. Obviously you don't need to include all of these, but there's a list of example user agents [here](http://www.useragentstring.com/pages/useragentstring.php).

**Comment by moonmeister** [comment]
@hmhealey I'll work on it. Thanks.

**Comment by moonmeister** [comment]
@hmhealey Okay, test's have been written. First time I've done this so let me know if anything needs to change. I made them fairly verbose.

**Comment by esethna** [comment]
Adding @lindy65 for help testing using browserstack

**Comment by esethna** [comment]
@hmhealey is going to help take over the review and testing

File: web/web.go [hunk 1]
```diff
@@ -46,17 +47,20 @@ func staticHandler(handler http.Handler) http.Handler {
 	})
 }
 
-var browsersNotSupported string = "MSIE/8;MSIE/9;MSIE/10;Internet Explorer/8;Internet Explorer/9;Internet Explorer/10;Safari/7;Safari/8"
+//List should be of minimum required browser version. Do not place a ; at the end of this string.
+var browsersNotSupported string = "MSIE/11;Internet Explorer/11;Safari/9;Chrome/43;Edge/15;Firefox/52"
```
Comment (crspeller) [inline web/web.go]: Can we declare this as a slice instead of string splitting it?

Comment (crspeller) [reply web/web.go]: Maybe a slice of structs for the fields? Avoid all the strange parsing?

Comment (moonmeister) [reply web/web.go]: @crspeller Yeah, I can look into it. I used what was already there, a slice makes more sense.

**Comment by moonmeister** [comment]
@crspeller I ended up using maps cause it allowed me to simplify the logic and eliminate the loop. Thanks for the suggestion, code seems much simpler.

**Comment by hmhealey** [comment]
Nice, that's much easier to tell which version of each is required

**Comment by jasonblais** [comment]
Looks like we'll need to do a dependency update for this PR: https://mattermost.atlassian.net/browse/PLT-7524

The dependancy update will be done for 4.3, so pushing this PR for 4.3 as well,

FIY @moonmeister @esethna

**Comment by esethna** [comment]
Ya thanks Jason, I think it's safer to let this soak given code complete is in 1 working day. Really appreciate all your work here @moonmeister, looking forward to getting this into master after the RC is cut!

**Comment by moonmeister** [comment]
@jasonblais @hmhealey @crspeller Can we get this merged before the repo split tonight since the RC has been cut?

**Comment by crspeller** [comment]
Yup let's get this merged. Good to go once the build passes.

**Pull request merged**